mono/sgen/sgen-marksweep.c

```diff
@@ -1818,6 +1818,9 @@ major_start_major_collection (void)
 			sweep_block (block);
 		SGEN_ASSERT (0, block->state == BLOCK_STATE_SWEPT, "All blocks must be swept when we're pinning.");
 		set_block_state (block, BLOCK_STATE_MARKING, BLOCK_STATE_SWEPT);
+		/* Blocks that have a null free_list are full. Evacuating them is pointless. */
```

Comment:
It's not necessarily pointless.  Until they're marked we don't know how many of those objects are live.